Introducing code generation library and SPI

pontusmelke opened the pull request:
-  Introduced code generation library
- All code generation operations are hidden behind SPI (still living in 2.3-compiler)
- `Instruction` use codegen library instead of producing strings
- Major clean up of the code generation code in `2.3-compiler`

pontusmelke committed:
Major clean up of Instruction and CodeGenerator

- Removed all unnused parts that was part of the old string generation.
- Now the `variables` map only contains `variables`
- Removed `JavaSymbol`
- ... and some overall clean up

thobe commented:
I would like to reorganise the commits of this. Hence marking it as `NOT READY FOR MERGE`, if possible we will update this PR, otherwise create a new one.

thobe committed:
Introduce code generation library

This code generation library builds on an abstraction close to Java (the language),
with methods representing the statements of Java that accept (a tree of) objects
representing the expressions of Java.

Currently there is one backend implemented for this code generation library, one
that generates java source code and compiles it through javac.

thobe committed:
Introduce Cypher SPI for code generation

With this SPI the Cypher compiler will know nothing about the underlying
methods it will be calling, only the operations it will want the generated
code to perform. The implementation of the SPI then knows about the specific
calls to make into the kernel API, generating code that makes those calls
directly.

The SPI is implemented through the use of the newly introduced code generation
library.

thobe commented:
Cleaned up commits, ready for review and merge (iff reviewer agrees).
